session.go

```diff
@@ -133,7 +139,10 @@ type session struct {
 
 	closeOnce sync.Once
 	// closeChan is used to notify the run loop that it should terminate
-	closeChan chan closeError
+	closeChan   chan closeError
+	delayedChan chan delayedError
+
+	prevState *recoveryState
```

Comment:
It seems like you don't need this here.  Instead, you can just have `saveState` return a `recoveryState`, and have `restoreState` take an argument.